pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/auth/FENCEAuthenticationService.java

```diff
@@ -132,24 +135,38 @@ private JsonNode getFENCEAccessToken(String callback_url, String fence_code) {
 
         String fence_token_url = JAXRSConfiguration.idp_provider_uri+"/user/oauth2/token";
 
+        // We have found that if the request is not quickly successful, it will time out.
+        // We are decreasing the timeout to 1 second to allow for quicker retries.
+        // We will fail after 3 retries.
         JsonNode resp = null;
-        try {
-            resp = HttpClientUtil.simplePost(
-                    fence_token_url,
-                    new StringEntity(query_string),
-                    JAXRSConfiguration.client,
-                    JAXRSConfiguration.objectMapper,
-                    headers.toArray(new Header[headers.size()])
-            );
-        } catch (Exception ex) {
-            logger.error("getFENCEAccessToken() failed to call FENCE token service, "+ex.getMessage());
+        try (CloseableHttpClient client = HttpClientBuilder.create().setDefaultRequestConfig(getFenceRequestConfig()).build()) {
+            int maxRetries = 3;
+            int retryCount = 0;
+            boolean success = false;
+            while (!success && retryCount < maxRetries) {
+                try {
+                    resp = HttpClientUtil.simplePost(
+                            fence_token_url,
+                            new StringEntity(query_string),
+                            client,
+                            JAXRSConfiguration.objectMapper,
+                            headers.toArray(new Header[headers.size()])
+                    );
+                    success = true;
+                } catch (Exception ex) {
+                    logger.error("getFENCEAccessToken() failed to call FENCE token service, {}", ex.getMessage());
+                    retryCount++;
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
         }
-        logger.debug("getFENCEAccessToken() finished: "+resp.asText());
+        logger.debug("getFENCEAccessToken() finished: {}", resp.asText());
```

Comment:
Are you logging a token here?

Reply:
I have to assume the resp.asText contains the user access token. We probably shouldn't be logging a token.